crates/plugins/src/manager.rs

```diff
@@ -0,0 +1,220 @@
+            true => {
+                fs::remove_file(manifest_file)?;
+                fs::remove_dir_all(plugin_store.plugin_subdirectory_path(plugin_name))?;
+                println!("Uninstalled plugin {} successfully", plugin_name);
+            }
+            false => {
+                println!(
+                    "Plugin \"{}\" isn't present, so no changes were made",
+                    plugin_name
+                );
+            }
+        }
+        Ok(())
+    }
+
+    /// Checks manifest to see if the plugin is compatible with the running version of Spin, does
+    /// not have a conflicting name with Spin internal commands, and is not a downgrade of a
+    /// currently installed plugin.
+    pub fn check_manifest(
+        &self,
+        plugin_manifest: &PluginManifest,
+        spin_version: &str,
+        allow_downgrades: bool,
+    ) -> Result<()> {
+        // Disallow installing plugins with the same name as spin internal subcommands
+        if SPIN_INTERNAL_COMMANDS
+            .iter()
+            .any(|&s| s == plugin_manifest.name())
+        {
+            bail!(
+                "Can't install a plugin with the same name ({}) as an internal plugin",
+                plugin_manifest.name()
+            );
+        }
+
+        // Disallow reinstalling identical plugins and downgrading unless permitted.
+        if let Ok(installed) = self.store.read_plugin_manifest(&plugin_manifest.name()) {
+            if &installed == plugin_manifest
+                || (installed.version > plugin_manifest.version && !allow_downgrades)
+            {
+                bail!(
+                    "Cannot install plugin {} {}. Plugin is already installed with version {}.",
+                    plugin_manifest.name(),
+                    plugin_manifest.version,
+                    installed.version,
+                );
+            }
+        }
+
+        check_supported_version(plugin_manifest, spin_version)
+    }
+
+    /// Fetches a manifest from a local, remote, or repository location and returned the parsed
+    /// PluginManifest object.
+    pub async fn get_manifest(
+        &self,
+        manifest_location: &ManifestLocation,
+    ) -> Result<PluginManifest> {
+        let plugin_manifest = match manifest_location {
+            ManifestLocation::Remote(url) => {
+                log::info!("Pulling manifest for plugin from {}", url);
+                reqwest::get(url.as_ref())
+                    .await?
+                    .json::<PluginManifest>()
+                    .await?
+            }
+            ManifestLocation::Local(path) => {
+                log::info!("Pulling manifest for plugin from {}", path.display());
+                let file = File::open(path).with_context(|| {
+                    format!("The local manifest {} could not be opened", path.display())
+                })?;
+                serde_json::from_reader(file)?
```

Comment:
Throughout this function it would be really really good to get some context on the various possible error returns (unless that's provided by whatever calls this, I'm not sure).  Like if JSON deserialisation fails, will the user have any more guidance than a serde "expected integer, got biscuits" kind of error?